File: APIs/mixitup-typescript-api/lib/currencies.ts
import { RestClient } from "./restclient";
import { ICurrency } from "./models/currency";
import { IUser } from "./models/user";
import { IGiveUserCurrency } from "./models/giveusercurrency";

export class Currencies {
    public static async getAllCurrenciesAsync(): Promise<ICurrency[]> {
        return await RestClient.GetAsync<ICurrency[]>("currency");
    }

    public static async getTopUsersByCurrencyAsync(currencyID: string, count?: number): Promise<IUser[]> {
        let uri: string = `currency/${currencyID}/top`;
        if (count !== undefined) {
            uri += `?count=${count}`;
        }

        return await RestClient.GetAsync<IUser[]>(uri);
    }

    public static async giveUsersCurrencyAsync(currencyID: string, giveList: IGiveUserCurrency[]): Promise<IUser[]> {
        if (giveList.length === 0) {
            return [];
        }

        return await RestClient.PostAsync<IUser[]>(`currency/${currencyID}/give`, giveList);
    }
}

File: APIs/MixItUp.API.NET8/V1/Currencies.cs
using MixItUp.API.V1.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace MixItUp.API
{
    //  GET:    http://localhost:8911/api/currency                      Currencies.GetAllCurrenciesAsync
    //  GET:    http://localhost:8911/api/currency/{currencyID}/top     Currencies.GetTopUsersByCurrencyAsync
    //  POST:   http://localhost:8911/api/currency/{currencyID}/give    Currencies.GiveUsersCurrencyAsync
    public static class Currencies
    {
        public static async Task<Currency[]> GetAllCurrenciesAsync()
        {
            return await RestClient.GetAsync<Currency[]>($"currency");
        }

        public static async Task<User[]> GetTopUsersByCurrencyAsync(Guid currencyID, uint? count)
        {
            string uri = $"currency/{currencyID}/top";
            if (count.HasValue)
            {
                uri += $"?count={count}";
            }

            return await RestClient.GetAsync<User[]>(uri);
        }

        public static async Task<User[]> GiveUsersCurrencyAsync(Guid currencyID, IEnumerable<GiveUserCurrency> giveList)
        {
            if (giveList.Count() == 0)
            {
                return new User[0];
            }

            return await RestClient.PostAsync<User[]>($"currency/{currencyID}/give", giveList);
        }
    }
}
